File: java/src/test/java/io/cucumber/gherkin/GherkinLineTest.java
package io.cucumber.gherkin;

import io.cucumber.messages.types.Location;
import org.junit.jupiter.api.Test;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import static java.util.Arrays.asList;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class GherkinLineTest {

    final Location line = Locations.atLine(12);

    @Test
    void allows_any_non_space_characters_in_a_tag() {
        GherkinLine gherkinLine = new GherkinLine("   @foo:bar  @zap\uD83E\uDD52yo", line);
        List<GherkinLineSpan> gherkinLineSpans = gherkinLine.parseTags();

        assertEquals(asList(
                new GherkinLineSpan(4, "@foo:bar"),
                new GherkinLineSpan(14, "@zap\uD83E\uDD52yo")
        ), gherkinLineSpans);
    }


    @Test
    void finds_tags() {
        GherkinLine gherkinLine = new GherkinLine("@this @is @a @tag", line);
        List<GherkinLineSpan> gherkinLineSpans = gherkinLine.parseTags();

        assertEquals(asList(
                new GherkinLineSpan(1, "@this"),
                new GherkinLineSpan(7, "@is"),
                new GherkinLineSpan(11, "@a"),
                new GherkinLineSpan(14, "@tag")
        ), gherkinLineSpans);
    }

    @Test
    void throws_on_tags_with_spaces() {
        GherkinLine gherkinLine = new GherkinLine("@this @is @a space separated @tag", line);
        assertThrows(ParserException.class, gherkinLine::parseTags);
    }

    @Test
    void throws_on_tags_with_leading_spaces() {
        GherkinLine gherkinLine = new GherkinLine("@ leadingSpace", line);
        assertThrows(ParserException.class, gherkinLine::parseTags);
    }

    @Test
    void ignores_empty_tag() {
        GherkinLine gherkinLine = new GherkinLine("@", line);
        List<GherkinLineSpan> gherkinLineSpans = gherkinLine.parseTags();

        assertEquals(Collections.emptyList(), gherkinLineSpans);
    }

    @Test
    void ignores_empty_tags() {
        GherkinLine gherkinLine = new GherkinLine("@@", line);
        List<GherkinLineSpan> gherkinLineSpans = gherkinLine.parseTags();

        assertEquals(Collections.emptyList(), gherkinLineSpans);
    }

    @Test
    void finds_tags__trim_whitespace() {
        GherkinLine gherkinLine = new GherkinLine("    @this @is  @a @tag  ", line);
        List<GherkinLineSpan> gherkinLineSpans = gherkinLine.parseTags();

        assertEquals(asList(
                new GherkinLineSpan(5, "@this"),
                new GherkinLineSpan(11, "@is"),
                new GherkinLineSpan(16, "@a"),
                new GherkinLineSpan(19, "@tag")
        ), gherkinLineSpans);
    }

    @Test
    void finds_tags__comment_after_tag() {
        GherkinLine gherkinLine = new GherkinLine("@this @is #acomment  ", line);
        List<GherkinLineSpan> gherkinLineSpans = gherkinLine.parseTags();

        assertEquals(asList(
                new GherkinLineSpan(1, "@this"),
                new GherkinLineSpan(7, "@is")
        ), gherkinLineSpans);
    }

    @Test
    void finds_tags__comment_inside_tag() {
        GherkinLine gherkinLine = new GherkinLine("@comment_tag#2 #a comment", line);
        List<GherkinLineSpan> gherkinLineSpans = gherkinLine.parseTags();

        assertEquals(asList(
                new GherkinLineSpan(1, "@comment_tag#2")
        ), gherkinLineSpans);
    }

    @Test
    void finds_tags__commented_before_tag() {
        GherkinLine gherkinLine = new GherkinLine("@this @is #@a commented tag", line);
        List<GherkinLineSpan> gherkinLineSpans = gherkinLine.parseTags();

        assertEquals(asList(
                new GherkinLineSpan(1, "@this"),
                new GherkinLineSpan(7, "@is")
        ), gherkinLineSpans);
    }

    @Test
    void finds_tags__commented_multiple_tags() {
        GherkinLine gherkinLine = new GherkinLine("@this @is #@a @commented @sequence of tags", line);
        List<GherkinLineSpan> gherkinLineSpans = gherkinLine.parseTags();

        assertEquals(asList(
                new GherkinLineSpan(1, "@this"),
                new GherkinLineSpan(7, "@is")
        ), gherkinLineSpans);
    }

    @Test
    void finds_tags_with_number() {
        GherkinLine gherkinLine = new GherkinLine("@ISSUE#123", line);
        List<GherkinLineSpan> gherkinLineSpans = gherkinLine.parseTags();

        assertEquals(asList(
                new GherkinLineSpan(1, "@ISSUE#123")
        ), gherkinLineSpans);
    }

    @Test
    void finds_table_cells() {
        // The cells below has the following whitespace characters on each side:
        // - U+00A0 (non-breaking space)
        // - U+0020 (space)
        // - U+0009 (tab)
        // This is generated with `ruby -e 'STDOUT.write "\u00A0\u0020\u0009".encode("utf-8")' | pbcopy`
        // and pasted.
        GherkinLine gherkinLine = new GherkinLine("      |  \tred  \t|  \tblue  \t|  \t\uD83E\uDD52\uD83E\uDD52\uD83E\uDD52  \t|  \tgreen  \t|", line);
        
        List<GherkinLineSpan> tableCells = gherkinLine.parseTableCells();
        GherkinLineSpan redSpan = tableCells.get(0);
        GherkinLineSpan blueSpan = tableCells.get(1);
        GherkinLineSpan emojiSpan = tableCells.get(2);
        GherkinLineSpan greenSpan = tableCells.get(3);

        assertEquals("red", redSpan.text);
        assertEquals(11, redSpan.column);

        assertEquals("blue", blueSpan.text);
        assertEquals(21, blueSpan.column);

        assertEquals("\uD83E\uDD52\uD83E\uDD52\uD83E\uDD52", emojiSpan.text);
        assertEquals(32, emojiSpan.column);

        assertEquals("green", greenSpan.text);
        assertEquals(42, greenSpan.column);
    }

    @Test
    void finds_escaped_table_cells() {
        GherkinLine gherkinLine = new GherkinLine("      | \\|æ\\\\n     | \\o\\no\\  | \\\\\\|a\\\\\\\\n | ø\\\\\\nø\\\\|", line);

        List<String> texts = gherkinLine.parseTableCells().stream().map(span -> span.text).collect(Collectors.toList());
        assertEquals(asList("|æ\\n", "\\o\no\\", "\\|a\\\\n", "ø\\\nø\\"), texts);
    }

    @Test
    void preserve_escaped_new_lines_at_start_and_end() {
        GherkinLine gherkinLine = new GherkinLine("      |  \nraindrops--\nher last kiss\ngoodbye.\n  |", line);
        List<String> texts = gherkinLine.parseTableCells().stream().map(span -> span.text).collect(Collectors.toList());
        assertEquals(asList("" +
                "\n" +
                "raindrops--\n" +
                "her last kiss\n" +
                "goodbye.\n"
        ), texts);
    }

    @Test
    void escapes_backslash() {
        GherkinLine gherkinLine = new GherkinLine("|\\\\o\\no\\||", line);
        List<String> texts = gherkinLine.parseTableCells().stream().map(span -> span.text).collect(Collectors.toList());
        assertEquals(asList("\\o\no|"), texts);
    }

    @Test
    void throws_on_illegal_escapes_backslash() {
        GherkinLine gherkinLine = new GherkinLine("|\\o\\no\\||", line);
        List<String> texts = gherkinLine.parseTableCells().stream().map(span -> span.text).collect(Collectors.toList());
        assertEquals(asList("\\o\no|"), texts);
    }

    @Test
    void correctly_trims_white_spaces_before_cell_content() {
        GherkinLine gherkinLine = new GherkinLine("|   \t spaces before|", line);
        List<String> texts = gherkinLine.parseTableCells().stream().map(span -> span.text).collect(Collectors.toList());
        assertEquals(asList("spaces before"), texts);
    }

    @Test
    void correctly_trims_white_spaces_after_cell_content() {
        GherkinLine gherkinLine = new GherkinLine("|spaces after   |", line);
        List<String> texts = gherkinLine.parseTableCells().stream().map(span -> span.text).collect(Collectors.toList());
        assertEquals(asList("spaces after"), texts);
    }

    @Test
    void correctly_trims_white_spaces_around_cell_content() {
        GherkinLine gherkinLine = new GherkinLine("|   \t spaces everywhere   \t|", line);
        List<String> texts = gherkinLine.parseTableCells().stream().map(span -> span.text).collect(Collectors.toList());
        assertEquals(asList("spaces everywhere"), texts);
    }

    @Test
    void does_not_drop_white_spaces_inside_a_cell() {
        GherkinLine gherkinLine = new GherkinLine("| foo()\n  bar\nbaz |", line);
        List<String> texts = gherkinLine.parseTableCells().stream().map(span -> span.text).collect(Collectors.toList());
        assertEquals(asList("foo()\n  bar\nbaz"), texts);
    }

    @Test
    void startsWithTitleKeyword_corresponding_keyword_match() {
        // Given
        GherkinLine gherkinLine = new GherkinLine("Feature: Hello", line);

        // When/Then
        assertTrue(gherkinLine.startsWithTitleKeyword("Feature"));
    }

    @Test
    void startsWithTitleKeyword_non_corresponding_keyword_does_not_match() {
        // Given
        GherkinLine gherkinLine = new GherkinLine("Rule: X", line);

        // When/Then
        assertFalse(gherkinLine.startsWithTitleKeyword("Background")); // not the same keyword
        assertFalse(gherkinLine.startsWithTitleKeyword("Rule: X")); // same keyword but with colon
        assertFalse(gherkinLine.startsWithTitleKeyword("Rul")); // shorter than keyword
    }

    @Test
    void startsWithTitleKeyword_keyword_without_colon_does_not_match() {
        // Given
        GherkinLine gherkinLine = new GherkinLine("Rule", line);

        // When/Then
        assertFalse(gherkinLine.startsWithTitleKeyword("Rule"));
    }

    @Test
    void parseTags_returns_empty_list_when_empty_line() {
        // Given
        GherkinLine gherkinLine = new GherkinLine("", line);

        // When/Then
        assertTrue(gherkinLine.parseTags().isEmpty());
    }

    @Test
    void substringTrimmed_empty_idempotence() {
        // Given
        GherkinLine gherkinLine = new GherkinLine("", line);

        // When/Then
        assertEquals("", gherkinLine.substringTrimmed(0));
    }

    @Test
    void substringTrimmed_spaces_after_index_are_removed() {
        // Given
        GherkinLine gherkinLine = new GherkinLine("Rule:    my rule name", line);

        // When/Then
        assertEquals("my rule name", gherkinLine.substringTrimmed(5));
    }

}

File: dart/test/src/language/gherkin_line_test.dart
import 'package:cucumber_gherkin/cucumber_gherkin.dart';
import 'package:cucumber_gherkin/src/language/gherkin_line.dart';
import 'package:cucumber_gherkin/src/language/gherkin_line_span.dart';
import 'package:test/test.dart';

void main() {
  const line = 12;

  test('Finds tags', () {
    final gherkinLine = GherkinLine('@this @is @a @tag', line);
    final gherkinLineSpans = gherkinLine.tags;

    expect(const [
      GherkinLineSpan(1, '@this'),
      GherkinLineSpan(7, '@is'),
      GherkinLineSpan(11, '@a'),
      GherkinLineSpan(14, '@tag'),
    ], gherkinLineSpans);
  });

  test('Throws on tags with spaces', () {
    final gherkinLine = GherkinLine('@this @is @a space separated @tag', line);
    expect(
      () => gherkinLine.tags,
      throwsA(const TypeMatcher<ParserException>()),
    );
  });

  test('Throws on tags with leading spaces', () {
    final gherkinLine = GherkinLine('@ leadingSpace', line);
    expect(
      () => gherkinLine.tags,
      throwsA(const TypeMatcher<ParserException>()),
    );
  });

  test('Ignores empty tag', () {
    final gherkinLine = GherkinLine('@', line);
    final gherkinLineSpans = gherkinLine.tags;
    expect(<GherkinLineSpan>[], gherkinLineSpans);
  });

  test('Ignores empty tags', () {
    final gherkinLine = GherkinLine('@@', line);
    final gherkinLineSpans = gherkinLine.tags;
    expect(<GherkinLineSpan>[], gherkinLineSpans);
  });

  test('Finds tags trim whitespace', () {
    final gherkinLine = GherkinLine('    @this @is  @a @tag  ', line);
    final gherkinLineSpans = gherkinLine.tags;
    expect(const [
      GherkinLineSpan(5, '@this'),
      GherkinLineSpan(11, '@is'),
      GherkinLineSpan(16, '@a'),
      GherkinLineSpan(19, '@tag'),
    ], gherkinLineSpans);
  });

  test('Finds tags comment inside tag', () {
    final gherkinLine = GherkinLine('@this @is #acomment  ', line);
    final gherkinLineSpans = gherkinLine.tags;

    expect(const [
      GherkinLineSpan(1, '@this'),
      GherkinLineSpan(7, '@is'),
    ], gherkinLineSpans);
  });

  test('Finds tags commented before tag', () {
    final gherkinLine = GherkinLine('@this @is #@a commented tag', line);
    final gherkinLineSpans = gherkinLine.tags;

    expect(const [
      GherkinLineSpan(1, '@this'),
      GherkinLineSpan(7, '@is'),
    ], gherkinLineSpans);
  });

  test('Finds tags commented multiple tags', () {
    final gherkinLine = GherkinLine(
      '@this @is #@a @commented @sequence of tags',
      line,
    );
    final gherkinLineSpans = gherkinLine.tags;

    expect(const [
      GherkinLineSpan(1, '@this'),
      GherkinLineSpan(7, '@is'),
    ], gherkinLineSpans);
  });

  test('finds_table_cells', () {
    // The cells below has the following whitespace characters on each side:
    // - U+00A0 (non-breaking space)
    // - U+0020 (space)
    // - U+0009 (tab)
    // The input also includes cucumber emoji cells to verify Unicode columns.
    final gherkinLine = GherkinLine(
      '      |  \tred  \t|  \tblue  \t|  \t'
      '\uD83E\uDD52\uD83E\uDD52\uD83E\uDD52  \t|  \tgreen  \t|',
      line,
    );
    final redSpan = gherkinLine.tableCells.first;
    final blueSpan = gherkinLine.tableCells.elementAt(1);
    final emojiSpan = gherkinLine.tableCells.elementAt(2);
    final greenSpan = gherkinLine.tableCells.elementAt(3);

    expect('red', redSpan.text);
    expect(11, redSpan.column);

    expect('blue', blueSpan.text);
    expect(21, blueSpan.column);

    expect('\uD83E\uDD52\uD83E\uDD52\uD83E\uDD52', emojiSpan.text);
    expect(32, emojiSpan.column);

    expect('green', greenSpan.text);
    expect(42, greenSpan.column);
  });

  test('counts_combining_marks_as_separate_columns', () {
    // The first cell holds "cafe" followed by a combining acute accent
    // (U+0301), written as base letter + combining mark. Columns
    // are counted per Unicode code point, so the combining mark advances the
    // column on its own and the next cell starts one column later than it would
    // under grapheme-cluster counting.
    final gherkinLine = GherkinLine('|cafe\u0301|test|', line);

    final firstCell = gherkinLine.tableCells.first;
    final secondCell = gherkinLine.tableCells.elementAt(1);

    expect('cafe\u0301', firstCell.text);
    expect(2, firstCell.column);

    expect('test', secondCell.text);
    // 8 with code-point counting; would be 7 if the combining sequence were
    // counted as a single grapheme cluster.
    expect(8, secondCell.column);
  });

  test('finds_escaped_table_cells', () {
    final gherkinLine = GherkinLine(
      '      | \\|\u00e6\\\\n     | \\o\\no\\  | '
      '\\\\\\|a\\\\\\\\n | \u00f8\\\\\\n\u00f8\\\\|',
      line,
    );

    final texts = gherkinLine.tableCells.map((span) => span.text).toList();
    expect(['|\u00e6\\n', '\\o\no\\', r'\|a\\n', '\u00f8\\\n\u00f8\\'], texts);
  });

  test('preserve_escaped_new_lines_at_start_and_end', () {
    final gherkinLine = GherkinLine(
      '      |  \nraindrops--\nher last kiss\ngoodbye.\n  |',
      line,
    );
    final texts = gherkinLine.tableCells.map((span) => span.text).toList();
    expect(['\nraindrops--\nher last kiss\ngoodbye.\n'], texts);
  });

  test('escapes_backslash', () {
    final gherkinLine = GherkinLine(r'|\\o\no\||', line);
    final texts = gherkinLine.tableCells.map((span) => span.text).toList();
    expect(['\\o\no|'], texts);
  });

  test('throws_on_illegal_escapes_backslash', () {
    final gherkinLine = GherkinLine(r'|\o\no\||', line);
    final texts = gherkinLine.tableCells.map((span) => span.text).toList();
    expect(['\\o\no|'], texts);
  });
}
